source/shared/ed25519-lc.c

```diff
@@ -0,0 +1,351 @@
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, NULL);
+    if (ctx == NULL) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        return NULL;
+    }
+
+    if (EVP_PKEY_keygen_init(ctx) <= 0) {
+        goto on_error;
+    }
+
+    if (EVP_PKEY_keygen(ctx, &pkey) <= 0) {
+        goto on_error;
+    }
+
+    struct aws_ed25519_key_pair *key_pair = aws_mem_calloc(allocator, 1, sizeof(struct aws_ed25519_key_pair));
+
+    aws_ref_count_init(&key_pair->ref_count, key_pair, s_ed25519_destroy_key);
+    key_pair->allocator = allocator;
+    key_pair->key = pkey;
+
+    return key_pair;
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    return NULL;
+#endif
+}
+
+struct aws_ed25519_key_pair *aws_ed25519_key_pair_acquire(struct aws_ed25519_key_pair *key_pair) {
+    return aws_ref_count_acquire(&key_pair->ref_count);
+}
+
+struct aws_ed25519_key_pair *aws_ed25519_key_pair_release(struct aws_ed25519_key_pair *key_pair) {
+    if (key_pair != NULL) {
+        aws_ref_count_release(&key_pair->ref_count);
+    }
+    return NULL;
+}
+
+static struct aws_byte_cursor s_key_type_literal = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("ssh-ed25519");
+
+int s_ed25519_openssh_encode_public_key(const struct aws_ed25519_key_pair *key_pair, struct aws_byte_buf *out) {
+    if (!aws_byte_buf_write_be32(out, s_key_type_literal.len) ||
+        aws_byte_buf_append(out, &s_key_type_literal) != AWS_OP_SUCCESS || !aws_byte_buf_write_be32(out, 32)) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+
+    size_t pub_len = 32;
+    AWS_FATAL_ASSERT(out->capacity - out->len >= pub_len);
+    if (EVP_PKEY_get_raw_public_key(key_pair->key, out->buffer + out->len, &pub_len) <= 0) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+    AWS_ASSERT(pub_len == 32);
+    out->len += 32;
+
+    return AWS_OP_SUCCESS;
+}
+
+/**
+ * format here is b64 of the following structure
+ * string "ssh-ed25519" #literal
+ * string key
+ * Note: string is always u32 size followed by the data. all multibyte ints are in big-endian
+ */
+int s_ed25519_export_public_openssh(const struct aws_ed25519_key_pair *key_pair, struct aws_byte_buf *out) {
+    uint8_t key_data[4 /*id len*/ + 11 /* ssh-ed25519 literal */ + 4 /*key len*/ + 32 /* key */] = {0};
+
+    struct aws_byte_buf key_buf = aws_byte_buf_from_empty_array(key_data, AWS_ARRAY_SIZE(key_data));
+
+    if (s_ed25519_openssh_encode_public_key(key_pair, &key_buf)) {
+        return AWS_OP_ERR;
+    }
+
+    struct aws_byte_cursor key_cur = aws_byte_cursor_from_buf(&key_buf);
+
+    if (aws_base64_encode(&key_cur, out) != AWS_OP_SUCCESS) {
+        return AWS_OP_ERR;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int s_ed25519_export_public_raw(const struct aws_ed25519_key_pair *key_pair, struct aws_byte_buf *out) {
+    size_t remaining = out->capacity - out->len;
+    if (remaining < 32) {
```

Comment:
debatable: should we be more flexible and call EVP_PKEY_get_raw_public_key() with NULL to learn the size?

Reply:
im not the biggest fan of calling it twice. we know the size must be 32 and we have buffer preallocated for that, if it wrote less bytes than that, we know something went wrong

Reply:
Pulling these into a constants might be helpful since we are using it multiple places.